### PR TITLE
split out NamedCType from CType. Remove direct string comparison from autograd codegen

### PR DESCRIPTION
**Pull request by bdhirsh**
The goal of this PR is to clean up some of the autograd codegen to compare C++ types using `CType` objects instead of raw strings. My last PR in the stack made that string comparison a little more fragile, since the raw C++ strings needed to be namespace-aware.

I confirmed byte-for-byte no codegen changes vs. the last PR (which added namespaces to the codegen) by running `diff -qr ../pytorch-common_test/torch/csrc/autograd/generated/ ../pytorch-callgrind_test_after2/torch/csrc/autograd/generated/` and `diff -qr ../pytorch-common_test/build/aten/src/ATen/ ../pytorch-callgrind_test_after2/build/aten/src/ATen/`

Note that a better end-state for the autograd codegen would be to do all of its type pattern matching directly off of JIT types, instead of off of CType’s (which are really just generated from JIT types, incorporating C++ specific semantics). That looks like it’ll require a pretty substantial change though, so I’m not doing it in this PR.

As part of this change (and after talking with @ezyang), I split off the `CType` data class into a separate `NamedCType` class, which holds a name and a `CType`. This way, `CType` only knows about actual C++ types, making it easier to compare CType’s to each other in the codegen when we only care about the type. The core change is in `types.py`, but it required a bunch of downstream changes to update all of the places where we create `CType`s to create `NamedCType`s instead.

The main change in the autograd codegen was that I updated `SavedAttribute` to store a `NamedCType`. The other autograd changes all pretty much came from that change.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55796 remove bridge API from codegen
* #55795 add _to_cpu() operator
* #55050 [EXPERIMENTAL, WIP] generate xla codegen in-tree
* **#55334 split out NamedCType from CType. Remove direct string comparison from autograd codegen**
* #55047 add C++ namespacing logic to ctypes
* #55046 Return a CType instead of a string for returns, beef up CType

Differential Revision: [D27708347](https://our.internmc.facebook.com/intern/diff/D27708347)

### CODE REVIEW

File: tools/autograd/load_derivatives.py [hunk 1]
```diff
@@ -251,91 +250,88 @@ def stride_expr(name: str) -> str:
...
-            formula = re.sub(regex.format(name), repl, formula)
+            formula = re.sub(regex.format(nctype.name), repl, formula)
 
         # Find any variables which remain in the formula and save them
-        if re.search(IDENT_REGEX.format(name), formula):
+        if re.search(IDENT_REGEX.format(nctype.name), formula):
             saved.append(SavedAttribute(
-                name=name,
-                # TODO: change from string to type data model
-                type=type.replace('const ', '').replace(' &', ''),
```
Comment (bdhirsh) [inline tools/autograd/load_derivatives.py:337]: I added a `remove_const_ref()` method to `CType` to emulate this in the data model.... it feels a little better, but it also required copying it into the 9 different `CType` classes we now have.

File: tools/codegen/api/translate.py [hunk 2]
```diff
@@ -115,38 +115,38 @@ def unsat(goal: CType) -> NoReturn:
...
 
         # const & is satisfied with mutable &
-        if isinstance(goal, ConstRefCType):
+        if isinstance(goal.type, ConstRefCType):
             try:
                 # WARNING: not strictly decreasing; be careful not
                 # to add a direct conversion that goes satisfies
                 # mutable& with const&
-                return solve(MutRefCType(goal.elem), direct=direct)
+                return solve(NamedCType(goal.name, MutRefCType(goal.type.elem)), direct=direct)
```
Comment (bdhirsh) [inline tools/codegen/api/translate.py:132]: This kind of thing is a downside of separating out `CType` and `NamedCType` - which Ed already pointed out. The "name" in `NamedCType` is only attached to the top-level CType. But it does make comparing C++ types easier :)

File: tools/autograd/gen_autograd_functions.py [hunk 3]
```diff
@@ -291,19 +291,21 @@ def process_function(info: DifferentiabilityInfo, template: CodeTemplate) -> str
...
 
     def save_var(var: SavedAttribute, is_output: bool) -> None:
-        name = var.name
+        name = var.nctype.name
+        type = var.nctype.type
         should_append_getsetdef = True
 
-        if var.type == 'at::Tensor' or var.type == 'c10::optional<at::Tensor>' or var.type == 'c10::optional<at::Tensor>&' or \
-                (var.type == 'at::Scalar' and is_output):
+        if type == BaseCType(tensorT) or type == OptionalCType(BaseCType(tensorT)) or \
```
Comment (ezyang) [inline tools/autograd/gen_autograd_functions.py:298]: maybe we should have some sort of set of `tensorBT` names for `BaseCType(tensorT)` ha!

**Review by ezyang — APPROVED** [review]
Good show, thanks. Curious how other people feel about removal of the C++ strings.

File: tools/codegen/api/types.py [hunk 4]
```diff
@@ -227,6 +211,36 @@ def name(self) -> ArgName:
...
+# A NamedCType is short for Named C++ semantic type.  A NamedCType represents a C++ type, plus
+# semantic information about what it represents.  For example, consider the
+# argument "bool pin_memory"; its normal C++ type is "bool", but its C++
+# semantic type also keeps track that this represents a "pin_memory"; you can't
+# just use a random other boolean in a context where you need a "pin_memory"!
+#
+# NamedCTypes encode C++ type structure as needed for translation.  Right now we
+# track references and optional, but don't, for example, track ArrayRef.  If
+# you need trnsnlations that know about these types, beef up this data
+# structure.
```
Comment (bhosmer) [inline tools/codegen/api/types.py:223]: I think this comment block actually still belongs with `CType`, no? (Also, the part about `ArrayRef` is out of date I think)

**Review by bhosmer — APPROVED** [review]

**Pull request closed without merging**